[WIP] Implement chat commands

dlqqq opened the pull request:
## Description

Implements a new system of chat commands in Jupyter Chat.

This PR is very much a work-in-progress. Right now, this branch provides a minimal working example of the design proposed in #144. CI is expected to fail for now.

## Demo

The chat commands API is now much more expressive and extensible. This design allows for multiple command providers, which can each provide a unique set of commands.

See this demo, where 2 custom providers are used to provide two sets of commands: one for slash commands, and one for emojis. Note that when an emoji command is invoked, it replaces the partial input with the emoji. This demonstrates how the new design can provide a more intuitive UX by allowing for operations previously unavailable in Jupyter Chat and Jupyter AI.

https://github.com/user-attachments/assets/073206cf-8a54-4889-af75-5c9d3e7657a5

jtpio commented on python/jupyterlab-chat/src/chat-commands/slash-command-demo.ts:
```diff
@@ -0,0 +1,58 @@
+export const slashCommandDemoPlugin: JupyterFrontEndPlugin<void> = {
+  id: 'jupyterlab-chat-extension:slashCommandDemoPlugin',
+  description:
+    'A demo plugin which adds Jupyter AI slash commands to the menu. Should be removed after PR review.',
```
Maybe this could be moved to the JupyterLite demo used in the docs?

https://github.com/jupyterlab/jupyter-chat/tree/main/docs/jupyter-chat-example

dlqqq replied:
It would make more sense to provide a demo after the `IChatCommandProvider` API is finalized. This would also be a great first issue for new contributors, so I'll open a new issue to see if we can get interest.

dlqqq replied:
Removed the demo. Opened a new issue for the Jupyterlite demo: #178.

brichet commented on packages/jupyter-chat/src/components/input/use-chat-commands.tsx:
```diff
@@ -0,0 +1,168 @@
+          return;
+        }
+
+        const replaceCurrentWordClosure = (newWord: string) => {
+          replaceCurrentWord(input, cursorIndex, newWord, setInput);
+        };
+
+        chatCommandRegistry.handleChatCommand(
+          command,
+          currentWord,
+          replaceCurrentWordClosure
+        );
```
Couldn't we replace the word in the `onChange()` callback directly. Since we have the command object, we could check if the attribute `replaceWith` is defined and replace the string. 

```ts
if (command.replaceWith) {
  replaceCurrentWord(input, cursorIndex, command.replaceWith, setInput);
}
```

Is there any advantage to replace it from the `handleChatCommand()` function ?

dlqqq replied:
Good thinking. I'll update the code & resolve this thread when I do so.

dlqqq replied:
Done!

brichet commented on packages/jupyter-chat/src/chat-commands/types.ts:
```diff
@@ -0,0 +1,70 @@
+   * TODO: Pass a ChatModel/InputModel instance here to provide a function to
+   * replace the current word.
+   */
+  handleChatCommand(
+    command: ChatCommand,
+    currentWord: string,
+    replaceCurrentWord: (newWord: string) => void
+  ): Promise<void>;
```
From the comment above, this interface would be reduced.
```suggestion
  handleChatCommand(
    command: ChatCommand
  ): Promise<void>;
```

but maybe I missed some use cases.

dlqqq replied:
Agreed, we should keep the interface simple. However, the `currentWord` and `replaceCurrentWord` params will be replaced with `chatModel/inputModel` once we start on #163. Therefore, I think we are free to ignore this concern for now.

dlqqq replied:
I've added #163 to the epic so we won't forget about this.

brichet commented on packages/jupyter-chat/src/components/input/use-chat-commands.tsx:
```diff
@@ -0,0 +1,168 @@
+        }
+    },
+    menu: {
+      open,
```
The `open` state is returned 2 times, in the `autocompleteProps` and in the `menu` props. Don't know if this expected, but both are used in the input component.

dlqqq replied:
I duplicated `open` because `chatCommands.menu.open` is more readable than `chatCommands.autocompleteProps.open`. I also found it helpful to distinguish the props being passed to the `Autocomplete` component (under `autocompleteProps`) from the state we are exposing to `ChatInput` (under `menu`).

We are still using a single source of truth (the `open` state variable), so I don't see a concern here. However, feel free to push changes to this branch if you think differently.

brichet commented on packages/jupyterlab-chat/src/factory.ts:
```diff
@@ -7,6 +7,7 @@ import {
   ChatWidget,
   IActiveCellManager,
   IAutocompletionRegistry,
```
We should probably remove all the references to `IAutocompletionRegistry`, `IAutocompletionCommandsProps` and maybe others, since this PR is a replacement to it.

dlqqq replied:
Agreed! This is one of the follow-up items I called out. I'll open a subissue for this.

brichet commented on python/jupyterlab-chat/src/chat-commands/registry.ts:
This file could live in `@jupyter/chat` package as a default registry that could be used by any extension, I think.

dlqqq replied:
Done. Please keep this conversation open for now so I will remember to open an issue.

brichet replied:
> BTW, how do you comment a file directly? I've never seen that in a GitHub code review 👀

The `comment` icon on the top right corner

![Screenshot from 2025-02-17 09-37-07](https://github.com/user-attachments/assets/8eeb607a-835d-487f-afc1-fec026abf7cc)

brichet replied:
> However, I have a feeling we could simplify this repo to just 2 packages: jupyterlab-chat should provide all the tokens, and jupyterlab-chat-extension should provide all the implementations.

The `jupyterlab-chat` provides the modules for a chat based on shared documents, while `@jupyter/chat` provides mostly the front-end components. Indeed, it also includes some token that can be useful for any chat using this library.

For me the question is most about the separated `jupyterlab-chat` and `jupyterlab-chat-extension` that could be reduce to one package. This separation is mostly to follow the pattern used in jupyterlab.

Feel free to open an issue about it if you think that it deserves more discussions.

dlqqq replied:
New issue: #173.

brichet reviewed (changes requested):
Thanks @dlqqq for working on it.

I left some comments below.

brichet commented:
In a more general test, I saw 2 issues in usages, that can be seen in the following screencast:
- after replacing an input text using a commands, a new line replace again the text with `undefined`
- the completion menu does not open if the word is not at the beginning of the input. Strangely, when printing the `autocompleteProps`, the `open` props is set to `true`, and the `options` array is not empty.

[record-2025-02-14_14.00.16.webm](https://github.com/user-attachments/assets/2b1b67e1-3087-45ce-b19c-043189b86458)

brichet commented on packages/jupyter-chat/src/components/input/use-chat-commands.tsx:
```diff
@@ -0,0 +1,168 @@
+
+      let newCommands: ChatCommand[] = [];
+      for (const provider of providers) {
+        // TODO: optimize performance when this method is truly async
```
Maybe this optimization could be perform at the `provider` level.

For example, it could keep in memory the previous word and suggestions. If the new word is only additional letters from the previous one, it only filters the previous suggestions.

dlqqq replied:
Agreed! Provider-level optimizations will be necessary for more complex command providers that need to make network requests frequently. However, what I meant by this comment was that we can also do some optimizations in our frontend:

- Use `Promise.all()` to start async `getChatCommands()` tasks in parallel.
- Show chat command suggestions as they resolve, instead of waiting for every provider to finish getting chat commands.

dlqqq replied:
Opened a new issue for this: https://github.com/jupyterlab/jupyter-chat/issues/167

dlqqq commented:
@brichet @jtpio Thank you both for the review. I've addressed your feedback and fixed the bugs that @brichet noticed.

I'll close the remaining threads once I finish opening issues. Unfortunately, I have something really urgent to attend to at work right now, so I'll finish that on Monday hopefully.

dlqqq commented:
Demo of latest branch state.

https://github.com/user-attachments/assets/1e4157e7-506c-4fe5-b21c-5ebfbb46a8da

brichet reviewed (commented):
Thanks @dlqqq for updating the PR, I can't reproduce the issues anymore :+1: .

dlqqq commented on packages/jupyter-chat/src/components/input/use-chat-commands.tsx:
```diff
@@ -0,0 +1,186 @@
+      return;
+    }
+
+    const replaceCurrentWordClosure = (newWord: string) => {
+      replaceCurrentWord(input, cursorIndex, newWord, setInput);
+    };
+
+    chatCommandRegistry.handleChatCommand(
+      command,
+      currentWord,
+      replaceCurrentWordClosure
+    );
```
I agree! I'm just keeping this here to indicate that we should implement a way of getting & setting the current word in the `InputModel` (#163). Some commands may need to replace the current word with something dynamic (i.e. not statically definable in `ChatCommand.replaceWith`).

dlqqq commented on python/jupyterlab-chat/src/chat-commands/registry.ts:
Done! 👍

brichet commented:
Thanks @dlqqq.
The failing test are related to the old autocompletion framework.
The best option would be to update them.
Otherwise we should at least skip them, and maybe add a fix to https://github.com/jupyterlab/jupyter-chat/issues/168 (or another issue).

dlqqq commented:
@brichet Since CI is already failing on `main`, can we go ahead and merge this while tracking the CI failure in a separate issue? Merging this now would make it a lot easier to work on #171.

dlqqq commented:
Finally finished opening all of the follow-up issues raised by other reviewers and me! Thank you all for taking the time to review this & ask questions. 🤗

jtpio commented:
> Since CI is already failing on `main`, can we go ahead and merge this while tracking the CI failure in a separate issue?

Not sure it's the same failure. Based on @brichet's comment above (https://github.com/jupyterlab/jupyter-chat/pull/161#issuecomment-2663606332) it seems related to the previous autocompletion registry implementation:

![image](https://github.com/user-attachments/assets/6b9d17fc-20d9-47cd-bbb5-ea9618f29d98)

Ideally we should be able to update these tests? If it's too much effort for now, maybe worth disabling / removing them for now, but track adding them back in an issue?

dlqqq commented:
@jtpio Thanks for the reminder! I've opened a new issue: #179

jtpio commented:
Nice thank! 

So let's remove or disable the outdated test code in this PR?

After https://github.com/jupyterlab/jupyter-chat/pull/148 the CI should normally be passing more consistently again (with skipped tests). And to lower maintenance burden it's nice to have a green check even though we know some tests have been disabled (it's easier to merge other PRs).

dlqqq commented:
@jtpio Good idea, I'll do that now.

dlqqq commented:
@jtpio  Done! 👍

brichet reviewed (approved):
Thanks @dlqqq and @jtpio
